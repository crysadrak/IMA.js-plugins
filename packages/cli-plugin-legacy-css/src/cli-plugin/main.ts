export type { LegacyCSSPluginOptions } from './LegacyCSSPlugin';
export { LegacyCSSPlugin } from './LegacyCSSPlugin';
