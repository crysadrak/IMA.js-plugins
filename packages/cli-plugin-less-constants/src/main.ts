export { LessConstantsPlugin } from './LessConstantsPlugin';

export * as units from './units/units';
export * as media from './units/media';
export { MapUnit, Unit, asUnit } from './units/utils';
