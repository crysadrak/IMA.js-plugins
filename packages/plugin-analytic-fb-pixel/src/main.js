import { pluginLoader } from '@ima/core';
import FacebookPixelAnalytic from './FacebookPixelAnalytic';

const defaultDependencies = FacebookPixelAnalytic.$dependencies;

pluginLoader.register('@ima/plugin-analytic-google', () => ({
  initSettings: () => ({
    prod: {
      plugin: {
        analytic: {
          fbPixel: {
            id: null
          }
        }
      }
    },

    test: {},

    dev: {}
  })
}));

export { FacebookPixelAnalytic, defaultDependencies };
