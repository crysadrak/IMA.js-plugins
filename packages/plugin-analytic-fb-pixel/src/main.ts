/* eslint-disable @typescript-eslint/no-empty-interface */
import { pluginLoader } from '@ima/core';

import FacebookPixelAnalytic from './FacebookPixelAnalytic';

declare module '@ima/core' {
  interface PluginSettings {}
  interface PluginAnalyticSettings {
    fbPixel: {
      id: string | null;
    };
  }

  interface Settings {
    plugin: PluginSettings;
  }
}

const defaultDependencies = FacebookPixelAnalytic.$dependencies;

pluginLoader.register('@ima/plugin-analytic-google', () => ({
  initSettings: () => ({
    prod: {
      plugin: {
        analytic: {
          fbPixel: {
            id: null,
          },
        },
      },
    },
  }),
}));

export { FacebookPixelAnalytic, defaultDependencies };
