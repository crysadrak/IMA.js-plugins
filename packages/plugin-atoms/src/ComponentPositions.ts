/* eslint-disable jsdoc/require-returns */
/* eslint-disable jsdoc/require-param */
import { ClientWindow, Dependencies } from '@ima/core';

export type VisibilityRect = {
  top: number;
  left: number;
  width: number;
  height: number;
};

/**
 * Component positions helper.
 */
export default class ComponentPositions {
  static $dependencies: Dependencies = ['$Window'];

  private _window: ClientWindow;

  /**
   * Initializes the helper.
   */
  constructor(window: ClientWindow) {
    this._window = window;
  }

  /**
   * Convert string to number.
   */
  convertToNumber(string: string) {
    try {
      return parseInt(string, 10);
    } catch (_) {
      /* empty */
    }

    return 0;
  }

  /**
   * Returns percent of visibility defined area in window viewport.
   */
  getPercentOfVisibility(elmRect: VisibilityRect) {
    if (!elmRect) {
      throw new Error(
        `Element rect is required. Call getBoundingClientRect()` +
          ` method on element or give object with properties { top: number,` +
          ` left: number, width: number, height: number }.`
      );
    }

    if (!this._window.isClient()) {
      return 0;
    }

    const intersectionRect = this.getRectsIntersection(
      this.getWindowViewportRect(),
      elmRect
    );

    return (
      ((intersectionRect.width * intersectionRect.height) /
        (elmRect.width * elmRect.height)) *
        100 || 0
    );
  }

  /**
   * Returns an intersection rectangle of two defined reactangles.
   */
  getRectsIntersection(
    rect1: VisibilityRect,
    rect2: VisibilityRect
  ): VisibilityRect {
    const top = this.getNumberFromRange(rect2.top, rect1.top, rect1.height);
    const left = this.getNumberFromRange(rect2.left, rect1.left, rect1.width);
    const bottom = this.getNumberFromRange(
      rect2.top + rect2.height,
      rect1.top,
      rect1.height
    );
    const right = this.getNumberFromRange(
      rect2.left + rect2.width,
      rect1.left,
      rect1.width
    );
    const width = right - left;
    const height = bottom - top;

    return { top, left, width, height };
  }

  /**
   * Returns number from defined range, if number is not in defined range return
   * min or max depends on number.
   */
  getNumberFromRange(number: number, min: number, max: number) {
    return Math.min(Math.max(number, min), max);
  }

  /**
   * Returns window viewport rect.
   */
  getWindowViewportRect(): VisibilityRect {
    const win = this._window.getWindow();
    const top = 0;
    const left = 0;
    let width = 0;
    let height = 0;

    if (this._window.isClient()) {
      width = win.innerWidth;
      height = win.innerHeight;
    }

    return { top, left, width, height };
  }

  /**
   * Returns window scroll position.
   */
  getWindowScrollPosition() {
    const left = window.scrollX || 0;
    const top = window.scrollX || 0;

    return { top, left };
  }

  /**
   * Returns the size of an element and its position relative to the viewport
   * and add extended value to returned rect.
   */
  getBoundingClientRect(
    element: HTMLElement,
    { width = 0, height = 0 }: { width?: number; height?: number },
    extended = 0
  ): VisibilityRect {
    if (!element || typeof element.getBoundingClientRect !== 'function') {
      throw new Error(
        `Element rect is required with callable getBoundingClientRect()` +
          ` method on element.`
      );
    }

    const clientRect = element.getBoundingClientRect();
    const isInvisibleElement =
      clientRect.top === 0 &&
      clientRect.left === 0 &&
      clientRect.width === 0 &&
      clientRect.height === 0;

    if (isInvisibleElement) {
      const { top, left, width, height } = clientRect;

      return { top, left, width, height };
    }

    const elmRectStyle = {
      top: clientRect.top - extended,
      left: clientRect.left - extended,
      width: clientRect.width + 2 * extended,
      height: (clientRect.height || height || 0 / width || 0) + 2 * extended,
    };

    return elmRectStyle;
  }
}
