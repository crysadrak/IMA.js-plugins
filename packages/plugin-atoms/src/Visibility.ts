/* eslint-disable jsdoc/require-param */
/* eslint-disable jsdoc/require-returns */
import {
  ClientWindow,
  Dependencies,
  Dispatcher,
  RouterEvents,
} from '@ima/core';
// @ts-expect-error
import { Circle } from 'infinite-circle';

type NotifyPayload = {
  type: string;
};

type NotifyCallback = (payload: NotifyPayload) => void;

/**
 * Visibility helper.
 */
export default class Visibility {
  static $dependencies: Dependencies = ['$Window', '$Dispatcher'];

  private window: ClientWindow;
  private dispatcher: Dispatcher;
  private afterHandleRouteCalled = false;
  circle: Circle;

  constructor(window: ClientWindow, dispatcher: Dispatcher) {
    this.window = window;
    this.dispatcher = dispatcher;
    this.circle = this.createVisibilityCircle();
  }

  /**
   * Create visibility circle.
   */
  private createVisibilityCircle() {
    return new Circle({
      listen: (notify: any) => this.listenOnEvents(notify),
      unlisten: (notify: any) => this.unlistenOnEvents(notify),
    });
  }

  /**
   * Register handlers to visibility loop
   */
  register(
    read: (...args: any[]) => any,
    write: (...args: any[]) => any,
    meta = { visibilityInterval: 180 }
  ): number {
    const id = this.circle.register({
      read,
      write,
      meta: { interval: meta.visibilityInterval },
    });

    if (this.afterHandleRouteCalled) {
      this.notify({ id });
    }

    return id;
  }

  /**
   * Unregister handlers from visibility loop
   */
  unregister(id: number) {
    this.circle.unregister(id);
  }

  /**
   * It cut down calling the event handler for defined interval. The throttle
   * method use requestAnimationFrame function which is called during page
   * scrolling.
   */
  throttle(
    eventHandler: (...args: any[]) => any,
    interval: number,
    context: any
  ) {
    const window = this.window.getWindow();
    interval = interval || 0;
    let callTime = 0;
    let lastArguments: any = null;

    if (context) {
      eventHandler = eventHandler.bind(context);
    }

    if (!this.window.isClient()) {
      return eventHandler;
    }

    function suspendAction() {
      if (callTime <= Date.now() || !window.requestAnimationFrame) {
        callTime = 0;
        eventHandler(...lastArguments);
      } else {
        window.requestAnimationFrame(suspendAction);
      }
    }

    return function throttle(...rest: any[]) {
      lastArguments = rest;

      if (!callTime) {
        callTime = Date.now() + interval;
        suspendAction();
      }
    };
  }

  /**
   * The method add circle instance to be running in the next infinite loop.
   */
  notify(...rest: any[]) {
    this.circle.notify(...rest);
  }

  /**
   * The visibility helper start checking visibility of registered entries.
   */
  private listenOnEvents(notify: NotifyCallback) {
    this.dispatcher.listen(
      RouterEvents.BEFORE_HANDLE_ROUTE,
      this.beforeHandleRoute,
      this
    );
    this.dispatcher.listen(
      RouterEvents.AFTER_HANDLE_ROUTE,
      this.afterHandleRoute,
      this
    );
    this.window.bindEventListener(this.window.getWindow(), 'resize', notify);
    this.window.bindEventListener(this.window.getWindow(), 'scroll', notify);
  }

  /**
   * The visibility helper stop checking visibility of registered entries.
   */
  private unlistenOnEvents(notify: NotifyCallback) {
    this.dispatcher.unlisten(
      RouterEvents.BEFORE_HANDLE_ROUTE,
      this.beforeHandleRoute,
      this
    );
    this.dispatcher.unlisten(
      RouterEvents.AFTER_HANDLE_ROUTE,
      this.afterHandleRoute,
      this
    );
    this.window.unbindEventListener(this.window.getWindow(), 'resize', notify);
    this.window.unbindEventListener(this.window.getWindow(), 'scroll', notify);
  }

  /**
   * The method resets `afterHandleRoute` marker.
   */
  private beforeHandleRoute() {
    this.afterHandleRouteCalled = false;
  }

  /**
   * The method normalize routeInfo to {@notifyPayload}.
   */
  private afterHandleRoute(routeInfo: any) {
    this.afterHandleRouteCalled = true;

    const payload = Object.assign(
      { type: RouterEvents.AFTER_HANDLE_ROUTE },
      routeInfo
    );

    this.notify(payload);
  }
}
