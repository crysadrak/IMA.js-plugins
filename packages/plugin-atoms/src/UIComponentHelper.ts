/* eslint-disable jsdoc/require-returns */
/* eslint-disable jsdoc/require-param */
import { Router, Window } from '@ima/core';
// @ts-expect-error
import { Infinite } from 'infinite-circle';

import ComponentPositions from './ComponentPositions';
import Visibility from './Visibility';

export type VisibilityOptions = {
  width?: number;
  height?: number;
  extendedPadding?: number;
  useIntersectionObserver: boolean;
  threshold?: number[];
};

/**
 * UI component helper.
 */
export default class UIComponentHelper {
  static $dependencies: [
    '$Router',
    '$Window',
    ComponentPositions,
    Visibility,
    Infinite,
    '$CssClasses'
  ];

  private router: Router;
  private window: Window;
  readonly componentPositions: ComponentPositions;
  readonly visibility: Visibility;
  readonly infinite: Infinite;
  private cssClassNameProcessor: (...args: any[]) => string; // TODO

  /**
   * Initializes the helper.
   */
  constructor(
    router: Router,
    window: Window,
    componentPositions: ComponentPositions,
    visibility: Visibility,
    infinite: Infinite,
    cssClassNameProcessor: (...args: any[]) => string // TODO
  ) {
    this.router = router;
    this.window = window;
    this.componentPositions = componentPositions;
    this.visibility = visibility;
    this.infinite = infinite;
    this.cssClassNameProcessor = cssClassNameProcessor;
  }

  /**
   * The public method which registers visibility circle to inifinite loop.
   */
  init() {
    this.infinite.add(this.visibility.circle);
  }

  /**
   * The regular expression was taken from the Closure sanitization library.
   */
  sanitizeUrl(url: string) {
    const SAFE_URL_PATTERN =
      /^(?:(?:https?|mailto|data|ftp|tel|file|sms):|[^&:/?#]*(?:[/?#]|$))/gi;

    url = String(url).trim();
    if (url.match(SAFE_URL_PATTERN)) return url;

    return 'unsafe:' + url;
  }

  /**
   * Serialize object as key and value pairs for using in noscript tag.
   */
  serializeObjectToNoScript(object: Record<string, any>) {
    return Object.keys(object).reduce((string, key) => {
      return string + ` ${key}="${object[key]}"`;
    }, '');
  }

  /**
   * Generate a string of CSS classes from the properties of the passed-in
   * object that resolve to true.
   */
  cssClasses(...classRuleGroups: (string | Record<string, boolean>)[]) {
    return this.cssClassNameProcessor(...classRuleGroups);
  }

  getVisibilityReader(element: HTMLElement, options: VisibilityOptions) {
    if (
      options.useIntersectionObserver &&
      // @ts-expect-error
      this.window.getWindow()?.IntersectionObserver
    ) {
      return this.getObserableReader(element, options);
    } else {
      return this.getReader(element, options);
    }
  }

  private getReader(element: HTMLElement, options: VisibilityOptions) {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const self = this;

    return function readVisibility() {
      const elementRect = self.componentPositions.getBoundingClientRect(
        element,
        { width: options.width, height: options.height },
        options.extendedPadding
      );

      return self.componentPositions.getPercentOfVisibility(elementRect);
    };
  }

  private getObserableReader(element: HTMLElement, options: VisibilityOptions) {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const self = this;
    const observerConfig = {
      rootMargin: options.extendedPadding + 'px',
      threshold: options.threshold || [0],
    };
    let intersectionObserverEntry: IntersectionObserverEntry | null = null;
    let isFirstPositionCalculated = false;

    const observer = new IntersectionObserver(entries => {
      intersectionObserverEntry = entries[0];
      this.visibility.circle.notify({ type: 'intersectionobserver', entries });
    }, observerConfig);
    observer.observe(element);

    return function readVisibility() {
      if (!isFirstPositionCalculated) {
        isFirstPositionCalculated = true;
        return { visibility: self.getReader(element, options)(), observer };
      }

      return { intersectionObserverEntry, observer };
    };
  }

  wrapVisibilityWriter(writer: (arg0: number, arg1?: any) => any) {
    return function parsePayload(circleEntry: any) {
      const { payload } = circleEntry;

      if (
        typeof payload === 'object' &&
        payload.observer &&
        payload.intersectionObserverEntry
      ) {
        const { intersectionObserverEntry: entry, observer } = payload;
        const isIntersectionBugged =
          entry.intersectionRatio === 0 && entry.isIntersecting;

        return writer(
          !isIntersectionBugged ? entry.intersectionRatio * 100 : 100,
          observer
        );
      } else if (
        typeof payload === 'object' &&
        payload.observer &&
        payload.visibility
      ) {
        return writer(payload.visibility, payload.observer);
      } else {
        return writer(payload);
      }
    };
  }
}
