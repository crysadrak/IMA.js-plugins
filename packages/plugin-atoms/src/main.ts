import './types';
import { ComponentUtils, pluginLoader } from '@ima/core';
import { Infinite } from 'infinite-circle';

import ComponentPositions from './ComponentPositions';
import { LAYOUT, LOADING } from './components/constants';
import {
  Headline,
  Headline1,
  Headline2,
  Headline3,
  Headline4,
  Headline5,
  Headline6,
} from './components/Headline';
import { Iframe } from './components/Iframe';
import { Image } from './components/Image';
import { Link } from './components/Link';
import { List, OrderedList, UnorderedList } from './components/List';
import { ListItem } from './components/ListItem';
import { Loader } from './components/Loader';
import { Paragraph } from './components/Paragraph';
import { Sizer } from './components/Sizer';
import UIComponentHelper, { type VisibilityOptions } from './UIComponentHelper';
import Visibility from './Visibility';

const defaultDependencies = ['$Router', ComponentPositions, Visibility];

pluginLoader.register('@ima/plugin-atoms', ns => {
  return {
    initBind: (ns, oc) => {
      oc.inject(Infinite, []);
    },
    initServices: (ns, oc) => {
      oc.get(UIComponentHelper).init();
      oc.get(ComponentUtils).register(
        '$UIComponentHelper',
        UIComponentHelper,
        '@ima/plugin-atoms'
      );
    },
  };
});

export type { VisibilityOptions };
export {
  UIComponentHelper,
  Visibility,
  ComponentPositions,
  defaultDependencies,
  Headline,
  Headline1,
  Headline2,
  Headline3,
  Headline4,
  Headline5,
  Headline6,
  Headline1 as H1,
  Headline2 as H2,
  Headline3 as H3,
  Headline4 as H4,
  Headline5 as H5,
  Headline6 as H6,
  Iframe,
  Image,
  Image as Img,
  Link,
  Link as A,
  List,
  ListItem,
  OrderedList,
  UnorderedList,
  ListItem as Li,
  OrderedList as Ol,
  UnorderedList as Ul,
  Loader,
  Paragraph,
  Paragraph as P,
  Sizer,
  LAYOUT,
  LOADING,
};
