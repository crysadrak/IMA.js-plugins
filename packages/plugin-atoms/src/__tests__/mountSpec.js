import { PageContext } from '@ima/react-page-renderer';
import classnames from 'classnames';
import { mount } from 'enzyme';
import { Infinite } from 'infinite-circle';
import { JSDOM } from 'jsdom';
import { toMockedInstance } from 'to-mock';

import _router from './mocks/router';
import _settings from './mocks/settings';
import _window from './mocks/window';
import ComponentPositions from '../ComponentPositions';
import * as UIAtoms from '../main';
import UIComponentHelper from '../UIComponentHelper';
// eslint-disable-next-line import/order
import Visibility from '../Visibility';

const jsdom = new JSDOM('<!doctype html><html><body></body></html>');
const { window } = jsdom;

global.window = window;
global.document = window.document;
global.navigator = {
  userAgent: 'node.js',
};

const visibility = toMockedInstance(Visibility);
const mockPosition = {
  height: 0,
};
const componentPositions = toMockedInstance(ComponentPositions, {
  getWindowViewportRect: () => mockPosition,
});
const infinite = toMockedInstance(Infinite);

function getComponentOptions(overrideSettings = {}) {
  const $Settings = Object.assign({}, _settings, overrideSettings);

  const $UIComponentHelper = new UIComponentHelper(
    _router,
    _window,
    componentPositions,
    visibility,
    infinite,
    classnames
  );

  const context = {
    $Utils: {
      $Settings,
      $UIComponentHelper,
      $CssClasses: classnames,
    },
  };
  const mountOptions = {
    context,
    someProp: true,
  };

  return mountOptions;
}

describe('UIAtoms mount rendering', () => {
  let wrapper = null;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
    }
  });

  describe('should render atoms without noscript tag: ', () => {
    let mountOptions = getComponentOptions();

    it('should render Image without noscript tag', () => {
      const Component = UIAtoms.Image;
      wrapper = mount(
        <PageContext.Provider value={mountOptions.context}>
          <Component src='example.jpg' />
        </PageContext.Provider>,
        mountOptions
      );

      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.find('noscript')).toHaveLength(0);
    });

    it('should render Iframe without noscript tag', () => {
      const Component = UIAtoms.Iframe;
      wrapper = mount(
        <PageContext.Provider value={mountOptions.context}>
          <Component src='example.html' />
        </PageContext.Provider>,
        mountOptions
      );

      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.find('noscript')).toHaveLength(0);
    });
  });
});
