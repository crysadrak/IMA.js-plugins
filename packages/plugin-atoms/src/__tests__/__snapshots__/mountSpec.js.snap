// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIAtoms mount rendering should render atoms without noscript tag:  should render Iframe without noscript tag 1`] = `"<div class="atm-iframe atm-overflow" style="width: auto; height: auto;"><iframe src="example.html" name="example.html" class="atm-fill"></iframe></div>"`;

exports[`UIAtoms mount rendering should render atoms without noscript tag:  should render Image without noscript tag 1`] = `"<img src="example.jpg" loading="lazy" decoding="async" class="atm-image atm-placeholder">"`;
